src/frontend/src/flows/authenticate/validateDerivationOrigin.ts

```diff
@@ -0,0 +1,76 @@
+import { Principal } from "@dfinity/principal";
+
+export type ValidationResult =
+  | { result: "valid" }
+  | { result: "invalid"; message: string };
+
+export const validateDerivationOrigin = async (
+  authRequestOrigin: string,
+  derivationOrigin?: string
+): Promise<ValidationResult> => {
+  if (
+    derivationOrigin === undefined ||
+    derivationOrigin === authRequestOrigin
+  ) {
+    // this is the default behaviour -> no further validation necessary
+    return { result: "valid" };
+  }
+
+  // check format of derivationOrigin
+  const matches = /^https:\/\/([\w-]*)(\.raw)?\.ic0\.app$/.exec(
+    derivationOrigin
+  );
+  if (matches === null) {
+    return {
+      result: "invalid",
+      message:
+        'derivationOrigin does not match regex "^https:\\/\\/([\\w-]*)(\\.raw)?\\.ic0\\.app$"',
+    };
+  }
+
+  try {
+    const canisterId = Principal.fromText(matches[1]); // verifies that a valid canister id was matched
+    const alternativeOriginsUrl = `https://${canisterId.toText()}.ic0.app/.well-known/ii-alternative-origins`;
+    const response = await fetch(
+      // always fetch non-raw
+      alternativeOriginsUrl,
+      // fail on redirects
+      { redirect: "error" }
+    );
+
+    if (!response.ok) {
```

Comment:
`response.ok` checks the 2xx range.

According to the spec only 200 is accepted. Could this be an issue?

Reply:
Hm, good point. I don't think this is a problem regarding security. But I'll restrict it anyway. 👍